File: Java/lf1/plp/functional1/Exemplo4.java
package lf1.plp.functional1;

import java.util.ArrayList;
import java.util.List;

import lf1.plp.functional1.declaration.DecFuncao;
import lf1.plp.functional1.declaration.DeclaracaoFuncional;
import lf1.plp.functional1.expression.Aplicacao;
import lf1.plp.functional1.expression.ExpDeclaracao;
import lf1.plp.functional1.expression.IfThenElse;
import lf1.plp.functional1.util.ValorFuncao;
import lf1.plp.expressions2.expression.ExpEquals;
import lf1.plp.expressions2.expression.ExpSoma;
import lf1.plp.expressions2.expression.ExpSub;
import lf1.plp.expressions2.expression.Expressao;
import lf1.plp.expressions2.expression.Id;
import lf1.plp.expressions2.expression.ValorInteiro;

/**
 * Avalia a expressao:
 *
 *   let fun mult x y = if (x == 0) then (0) else (y + mult(x - 1, y)) in
 *     mult(3, 4)
 *
 * Calculo:
 *   mult(3,4) = 4 + mult(2,4)
 *             = 4 + 4 + mult(1,4)
 *             = 4 + 4 + 4 + mult(0,4)
 *             = 4 + 4 + 4 + 0
 * Resultado esperado: 12
 */
public class Exemplo4 {

    static void executar(String descricao, Programa prog) {
        System.out.println("--------------------------------------------------");
        System.out.println("Expressao : " + descricao);
        try {
            boolean bemTipada = prog.checaTipo();
            System.out.println("Bem tipada: " + bemTipada);
            if (bemTipada) {
                System.out.println("Resultado : " + prog.executar());
            } else {
                System.out.println("Erro de tipo: avaliacao abortada.");
            }
        } catch (Exception e) {
            System.out.println("Excecao capturada: " + e.getMessage());
            e.printStackTrace();
        }
        System.out.println("--------------------------------------------------");
    }

    public static void main(String[] args) {

        // ------------------------------------------------------------------
        // Parametros formais de mult: x e y
        // ------------------------------------------------------------------
        Id idX = new Id("x");
        Id idY = new Id("y");

        List<Id> params = new ArrayList<Id>();
        params.add(idX);
        params.add(idY);

        // ------------------------------------------------------------------
        // Corpo recursivo: y + mult(x - 1, y)
        // ------------------------------------------------------------------

        // Argumentos da chamada recursiva: (x - 1)  e  y
        List<Expressao> argsRec = new ArrayList<Expressao>();
        argsRec.add(new ExpSub(new Id("x"), new ValorInteiro(1)));
        argsRec.add(new Id("y"));

        // Chamada recursiva: mult(x - 1, y)
        Expressao chamadaRec = new Aplicacao(new Id("mult"), argsRec);

        // Ramo else: y + mult(x - 1, y)
        Expressao ramoElse = new ExpSoma(new Id("y"), chamadaRec);

        // ------------------------------------------------------------------
        // Corpo completo de mult:
        //   if (x == 0) then (0) else (y + mult(x - 1, y))
        // ------------------------------------------------------------------
        Expressao condicao  = new ExpEquals(new Id("x"), new ValorInteiro(0));
        Expressao ramoThen  = new ValorInteiro(0);
        Expressao corpoMult = new IfThenElse(condicao, ramoThen, ramoElse);

        // ------------------------------------------------------------------
        // Declaracao: fun mult x y = <corpoMult>
        // ------------------------------------------------------------------
        ValorFuncao vfMult   = new ValorFuncao(params, corpoMult);
        DecFuncao   decMult  = new DecFuncao(new Id("mult"), vfMult);

        // ------------------------------------------------------------------
        // Corpo do let: mult(3, 4)
        // ------------------------------------------------------------------
        List<Expressao> argsMain = new ArrayList<Expressao>();
        argsMain.add(new ValorInteiro(3));
        argsMain.add(new ValorInteiro(4));
        Expressao aplicacaoMult = new Aplicacao(new Id("mult"), argsMain);

        // ------------------------------------------------------------------
        // Monta: let fun mult x y = ... in mult(3, 4)
        // ------------------------------------------------------------------
        List<DeclaracaoFuncional> decs = new ArrayList<DeclaracaoFuncional>();
        decs.add(decMult);
        Expressao letMult = new ExpDeclaracao(decs, aplicacaoMult);

        // ------------------------------------------------------------------
        // Executa o programa
        // ------------------------------------------------------------------
        Programa prog = new Programa(letMult);

        executar(
            "let fun mult x y = if (x == 0) then (0) else (y + mult(x - 1, y)) in\n" +
            "  mult(3, 4)",
            prog
        );
    }
}

File: Python/lf1-python/Exemplo4.py
"""
Avalia a expressao:

  let fun mult x y = if (x == 0) then (0) else (y + mult(x - 1, y)) in
    mult(3, 4)

Calculo:
  mult(3,4) = 4 + mult(2,4)
            = 4 + 4 + mult(1,4)
            = 4 + 4 + 4 + mult(0,4)
            = 4 + 4 + 4 + 0
Resultado esperado: 12
"""

from Programa import Programa
from declaration.dec_funcao import DecFuncao
from expression.exp_declaracao import ExpDeclaracao
from expression.aplicacao import Aplicacao
from expression.if_then_else import IfThenElse
from util.valor_funcao import ValorFuncao
from expression.exp_equals import ExpEquals
from expression.exp_soma import ExpSoma
from expression.exp_sub import ExpSub
from expression.id import Id
from expression.valor_inteiro import ValorInteiro


def executar(descricao: str, prog: Programa) -> None:
    print("--------------------------------------------------")
    print("Expressao :\n" + descricao)
    try:
        bem_tipada = prog.checa_tipo()
        print("Bem tipada: " + str(bem_tipada).lower())
        if bem_tipada:
            print("Resultado: " + str(prog.executar()))
        else:
            print("Erro de tipo: avaliacao abortada.")
    except Exception as e:
        print("Excecao capturada: " + str(e))
        import traceback
        traceback.print_exc()
    print("--------------------------------------------------")


def main():
    # Parametros formais de mult: x e y
    params = [Id("x"), Id("y")]

    # Corpo recursivo: y + mult(x - 1, y)
    args_rec = [ExpSub(Id("x"), ValorInteiro(1)), Id("y")]
    chamada_rec = Aplicacao(Id("mult"), args_rec)
    ramo_else = ExpSoma(Id("y"), chamada_rec)

    # Corpo completo de mult:
    #   if (x == 0) then (0) else (y + mult(x - 1, y))
    condicao = ExpEquals(Id("x"), ValorInteiro(0))
    ramo_then = ValorInteiro(0)
    corpo_mult = IfThenElse(condicao, ramo_then, ramo_else)

    # Declaracao: fun mult x y = <corpoMult>
    vf_mult = ValorFuncao(params, corpo_mult)
    dec_mult = DecFuncao(Id("mult"), vf_mult)

    # Corpo do let: mult(3, 4)
    args_main = [ValorInteiro(3), ValorInteiro(4)]
    aplicacao_mult = Aplicacao(Id("mult"), args_main)

    # Monta: let fun mult x y = ... in mult(3, 4)
    let_mult = ExpDeclaracao([dec_mult], aplicacao_mult)

    # Executa o programa
    prog = Programa(let_mult)

    executar(
        "let fun mult x y = if (x == 0) then (0) else (y + mult(x - 1, y)) in\n" +
        "  mult(3, 4)",
        prog
    )


if __name__ == "__main__":
    main()
